buildscripts/generate_macros.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/python
+
+# Copyright 2016 MongoDB Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Auto-generates the given number of levels of the MONGO_ODM_CHILD* macro.
+# These are placed in a file with the proper mongo-cxx-odm boilerplate.
+# This is used to to specify nested fields in the query builder.
+# BSON supports up to 100 levels of nesting.
+
+import sys
+
+def usage():
+    print("Usage: python generate_macros.py <outfile> [max_depth]")
+    print("Generates 'max_depth' levels of the MONGO_ODM_CHILD* macro, and write the output ")
+    print("to 'outfile'. ")
+    exit()
+
+def print_macros(outfile, max):
+    """
+    Prints the MONGO_ODM_CHILD* macros for *=2 to `max`.
+    :param max: The maximum level to which the macro is generated.
+    """
+
+    outfile.write(
+    """// clang-format off
+    // Copyright 2016 MongoDB Inc.
+    //
+    // Licensed under the Apache License, Version 2.0 (the "License");
+    // you may not use this file except in compliance with the License.
+    // You may obtain a copy of the License at
+    //
+    // http://www.apache.org/licenses/LICENSE-2.0
+    //
+    // Unless required by applicable law or agreed to in writing, software
+    // distributed under the License is distributed on an "AS IS" BASIS,
+    // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    // See the License for the specific language governing permissions and
+    // limitations under the License.
+
+    #pragma once
+
+    #include <mongo_odm/config/prelude.hpp>
+
+    """)
+
+
+    for i in range(2, max+1):
+        prev_fields = ", ".join(["field" + str(n) for n in range(1, i)])
+        all_fields = prev_fields + ", field" + str(i)
+        s = ("#define MONGO_ODM_CHILD%d(base, %s) "
+            "make_nvp_with_parent(MONGO_ODM_KEY_BY_VALUE(&std::decay_t<"
+            "typename decltype(MONGO_ODM_CHILD%d(base, %s))::type>::field%d), "
+            "MONGO_ODM_CHILD%d(base, %s))") % \
+            (i, all_fields, i-1, prev_fields, i, i-1, prev_fields)
+        outfile.write(s)
+        outfile.write("\n\n")
+
+    outfile.write(
+    """
+    #define MONGO_ODM_CHILD(type, ...) \
```

Comment:
I think `MONGO_ODM_CHILD` and `MONGO_ODM_CHILD_ELEM` should go into macros.hpp and `MONGO_ODM_CHILD1` should be in this one.